File: Backend/routes/ai.js
const express = require('express');
const { generateContent } = require('../services/geminiService');
const {
  buildResumePrompt,
  buildCoverLetterPrompt,
  buildRewritePrompt,
} = require('../services/promptBuilder');

module.exports = function (pool) {
  const router = express.Router();

  // POST /api/ai/resume/:email/:jobId
  // Generates a resume tailored to one specific job, using the user's
  // saved profile (summary/skills/education/experience) + that job's
  // description.
  router.post('/resume/:email/:jobId', async (req, res) => {
    try {
      const { email, jobId } = req.params;

      const profileResult = await pool.query(
        `SELECT first_name, last_name, summary, skills, education, experience, target_role
         FROM user_profile
         WHERE email = $1`,
        [email]
      );

      if (profileResult.rows.length === 0) {
        return res.status(404).json({
          success: false,
          error: 'Profile not found. Fill out your profile before generating a resume.',
        });
      }

      const jobResult = await pool.query(
        `SELECT title, company, description
         FROM job_table
         WHERE unique_num = $1 AND email = $2 AND is_deleted = FALSE`,
        [jobId, email]
      );

      if (jobResult.rows.length === 0) {
        return res.status(404).json({ success: false, error: 'Job not found.' });
      }

      const p = profileResult.rows[0];
      const job = jobResult.rows[0];

      const profileSummary = [
        `${p.first_name ?? ''} ${p.last_name ?? ''}`.trim(),
        p.target_role ? `Target role: ${p.target_role}` : null,
        p.summary ?? null,
      ]
        .filter(Boolean)
        .join('\n');

      const prompt = buildResumePrompt({
        profile: profileSummary,
        experience: p.experience,
        education: p.education,
        skills: p.skills,
        job: `${job.title} at ${job.company}\n\n${job.description}`,
      });

      const result = await generateContent(prompt);

      if (!result.success) {
        return res.status(502).json({ success: false, error: result.error });
      }

      res.json({
        success: true,
        content: result.content,
        job: { id: Number(jobId), title: job.title, company: job.company },
      });
    } catch (err) {
      console.error('AI resume generation error:', err);
      res.status(500).json({ success: false, error: 'Failed to generate resume.' });
    }
  });

  // POST /api/ai/edit
  // Rewrites/improves whatever document content is currently in the editor.
  // Body: { content, rewriteType?, jobDescription?, documentType? }
  // documentType defaults to 'Resume' so existing callers keep working, but
  // the Cover Letter editor passes documentType: 'Cover Letter'.
  router.post('/edit', async (req, res) => {
    try {
      const { content, rewriteType, jobDescription, documentType } =
        req.body ?? {};

      if (!content || !String(content).trim()) {
        return res
          .status(400)
          .json({ success: false, error: 'content is required.' });
      }

      const prompt = buildRewritePrompt({
        documentType: documentType?.trim() || 'Resume',
        rewriteType: rewriteType?.trim() || 'Professional, ATS-friendly',
        content,
        job: jobDescription ?? '',
      });

      const result = await generateContent(prompt);

      if (!result.success) {
        return res.status(502).json({ success: false, error: result.error });
      }

      res.json({ success: true, content: result.content });
    } catch (err) {
      console.error('AI edit error:', err);
      res
        .status(500)
        .json({ success: false, error: 'Failed to edit document.' });
    }
  });

  // POST /api/ai/cover-letter/:email/:jobId
  // Generates a cover letter tailored to one specific job, using the user's
  // saved profile (summary/skills/education/experience) + that job's
  // description.
  router.post('/cover-letter/:email/:jobId', async (req, res) => {
    try {
      const { email, jobId } = req.params;

      const profileResult = await pool.query(
        `SELECT first_name, last_name, summary, skills, education, experience, target_role
         FROM user_profile
         WHERE email = $1`,
        [email]
      );

      if (profileResult.rows.length === 0) {
        return res.status(404).json({
          success: false,
          error: 'Profile not found. Fill out your profile before generating a cover letter.',
        });
      }

      const jobResult = await pool.query(
        `SELECT title, company, description
         FROM job_table
         WHERE unique_num = $1 AND email = $2 AND is_deleted = FALSE`,
        [jobId, email]
      );

      if (jobResult.rows.length === 0) {
        return res.status(404).json({ success: false, error: 'Job not found.' });
      }

      const p = profileResult.rows[0];
      const job = jobResult.rows[0];

      const profileSummary = [
        `${p.first_name ?? ''} ${p.last_name ?? ''}`.trim(),
        p.target_role ? `Target role: ${p.target_role}` : null,
        p.summary ?? null,
      ]
        .filter(Boolean)
        .join('\n');

      const prompt = buildCoverLetterPrompt({
        profile: profileSummary,
        experience: p.experience,
        education: p.education,
        skills: p.skills,
        job: `${job.title} at ${job.company}\n\n${job.description}`,
      });

      const result = await generateContent(prompt);

      if (!result.success) {
        return res.status(502).json({ success: false, error: result.error });
      }

      res.json({
        success: true,
        content: result.content,
        job: { id: Number(jobId), title: job.title, company: job.company },
      });
    } catch (err) {
      console.error('AI cover letter generation error:', err);
      res
        .status(500)
        .json({ success: false, error: 'Failed to generate cover letter.' });
    }
  });

  return router;
};

File: Backend/routes/Ai.py
from flask import Blueprint
from flask import jsonify
from flask import request

from Models.User import User
from Models.Job import Job

from Services.GeminiService import GeminiService
from Services.PromptBuilder import PromptBuilder


aiBlueprint = Blueprint(
    "Ai",
    __name__
)

geminiService = GeminiService()

promptBuilder = PromptBuilder()


@aiBlueprint.route("/resume", methods=["POST"])
def GenerateResume():

    data = request.get_json()

    if data is None:

        return jsonify({

            "success": False,

            "error": "Request body is missing."

        }), 400

    userId = data.get("userId")
    jobId = data.get("jobId")

    if userId is None or jobId is None:

        return jsonify({

            "success": False,

            "error": "userId and jobId are required."

        }), 400

    user = User.query.get(userId)

    job = Job.query.get(jobId)

    if user is None:

        return jsonify({

            "success": False,

            "error": "User not found."

        }), 404

    if job is None:

        return jsonify({

            "success": False,

            "error": "Job not found."

        }), 404

    #
    # Replace these with your actual database fields
    #

    profile = user.profile

    experience = user.experience

    education = user.education

    skills = user.skills

    jobDescription = job.description

    prompt = promptBuilder.BuildResumePrompt(

        profile,

        experience,

        education,

        skills,

        jobDescription

    )

    result = geminiService.GenerateResume(prompt)

    return jsonify(result)


@aiBlueprint.route("/coverletter", methods=["POST"])
def GenerateCoverLetter():

    data = request.get_json()

    if data is None:

        return jsonify({

            "success": False,

            "error": "Request body is missing."

        }), 400

    userId = data.get("userId")
    jobId = data.get("jobId")

    user = User.query.get(userId)

    job = Job.query.get(jobId)

    if user is None or job is None:

        return jsonify({

            "success": False,

            "error": "User or Job not found."

        }), 404

    profile = user.profile

    experience = user.experience

    education = user.education

    skills = user.skills

    jobDescription = job.description

    prompt = promptBuilder.BuildCoverLetterPrompt(

        profile,

        experience,

        education,

        skills,

        jobDescription

    )

    result = geminiService.GenerateCoverLetter(prompt)

    return jsonify(result)


@aiBlueprint.route("/rewrite", methods=["POST"])
def RewriteDocument():

    data = request.get_json()

    if data is None:

        return jsonify({

            "success": False,

            "error": "Request body is missing."

        }), 400

    prompt = promptBuilder.BuildRewritePrompt(

        documentType=data.get("documentType"),

        rewriteType=data.get("rewriteType"),

        content=data.get("content"),

        job=data.get("job", "")

    )

    result = geminiService.RewriteDocument(prompt)

    return jsonify(result)


@aiBlueprint.route(
    "/company-research",
    methods=["POST"]
)
def CompanyResearch():

    data = request.get_json()

    company = data.get("company", "").strip()

    context = data.get("context", "").strip()

    if company == "":

        return jsonify({
            "success": False,
            "message": "Company name is required."
        }), 400

    promptBuilder = PromptBuilder()

    prompt = promptBuilder.BuildCompanyResearchPrompt(
        company,
        context
    )

    geminiService = GeminiService()

    result = geminiService.GenerateCompanyResearch(
        prompt
    )

    if not result["success"]:

        return jsonify(result), 500

    return jsonify(result), 200